supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,462 @@
+
+        if not len(completed_samples) == len(self.samples):
+            print("Some samples do not have probability results.")
+            if not force:
+                return False
+
+        if len(completed_samples) == 0:
+            raise RuntimeError("Cannot force data collection when there are no completed samples.")
+
+        self._raw_data = self.process_probabilities(completed_samples)
+        return True
+
+    @abstractmethod
+    def build_circuits(
+        self,
+        num_circuits: int,
+        layers: Iterable[int],
+    ) -> Sequence[Sample]:
+        """Build a list of circuits required for the experiment. These circuits are stored in
+        :class:`Sample` objects along with any additional data that is needed during the analysis.
+
+        Args:
+            num_circuits: Number of circuits to generate.
+            layers: An iterable of the different layer depths to use during the experiment.
+
+        Returns:
+            The list of circuit objects
+        """
+
+    @abstractmethod
+    def process_probabilities(self, samples: Sequence[Sample]) -> pd.DataFrame:
+        """Processes the probabilities generated by sampling the circuits into a data frame
+        needed for analyzing the results.
+        """
+
+    @abstractmethod
+    def plot_results(self) -> None:
+        """Plot the results of the experiment"""
+
+    @abstractmethod
+    def analyse_results(self, plot_results: bool = True) -> NamedTuple:
+        """Perform the experiment analysis and store the results in the `results` attribute"""
+
+    def _state_probs_to_dict(
+        self, probs: np.typing.NDArray[np.float64], prefix: str = "", suffix: str = ""
+    ) -> dict[str, float]:
+        """Converts a numpy array of state probabilities to a dictionary indexed
+        by the bitstring. Optional prefix and suffix can be added.
+
+        Args:
+            probs: Numpy array of coefficients.
+            prefix: Optional prefix to the bitstring key. Defaults to "".
+            suffix: Optional suffix to the bitstring key. Defaults to "".
+
+        Returns:
+            Dictionary of state probabilities indexed by bitstring.
+        """
+        return {
+            prefix + format(idx, f"0{self.num_qubits}b") + suffix: coefficient
+            for idx, coefficient in enumerate(probs)
+        }
+
+    @staticmethod
+    def _interleave_gate(
+        circuit: cirq.Circuit, gate: cirq.Gate, include_final: bool = False
+    ) -> cirq.Circuit:
+        """Interleave a given gate into a circuit.
+
+        Args:
+            circuit: The original circuit.
+            gate: The gate to interleave.
+            include_final: If True then the interleaving gate is also appended to
+                the end of the circuit.
+
+        Returns:
+            A copy of the original circuit with the provided gate interleaved.
+        """
+        qubits = circuit.all_qubits()
+        interleaved_circuit = deepcopy(circuit)
```

Comment:
I think `cirq.Circuit.copy` is `deep`:
```suggestion
        interleaved_circuit = circuit.copy()
```